app/com/yetu/oauth2provider/services/data/memory/MemoryPersonService.scala

```diff
@@ -1,104 +1,150 @@
-      ) yield {
-        identityWithPasswordInfo.passwordInfo.get
-      }
-    }
+    Future.successful(usersIds.values.find(_ == user).map(u => u.passwordInfo.get))
   }
 
-  def find(providerId: String, userId: String): Future[Option[BasicProfile]] = {
-    Future.successful(findYetuUser(userId).map(_.toBasicProfile))
+  def find(providerId: String, userId: String) = {
+    findUser(userId).map {
+
+      case Some(user) =>
+        if (user.providerId.equals(providerId)) {
+          Some(user)
+        } else None
+
+      case _ => None
+    }
   }
 
   def findByEmailAndProvider(email: String, providerId: String): Future[Option[BasicProfile]] = {
-    find(providerId, email)
+
+    val user = usersEmails.find(_._1 == email).map(_._2)
+    val result = user match {
+
+      case Some(u) =>
+        if (u.providerId.equals(providerId)) {
+          Some(u)
+        } else None
+
+      case _ => None
+    }
+
+    Future.successful(result)
   }
 
-  def save(user: BasicProfile, mode: SaveMode): Future[YetuUser] = {
+  override def save(user: BasicProfile, mode: SaveMode) = {
     logger.debug(s"Save user $user")
-    Future.successful {
-      val userToReturn: YetuUser = mode match {
-        case SaveMode.SignUp => {
-          val newUser = YetuUserHelper.fromBasicProfile(user, UUIDGenerator.uuid())
-          logger.debug(s"saving user $newUser")
-          addNewUser(newUser)
-          newUser
-        }
-        case SaveMode.PasswordChange => {
-          val oldUser = findYetuUser(user.userId).get
-          val newUser = oldUser.copy(passwordInfo = user.passwordInfo)
-          users += user.userId -> newUser
-          findYetuUser(user.userId).get
-        }
-        case _ =>
-          logger.warn("not saving as signUp; ignoring request.")
-          findYetuUser(user.userId).get
+    val result = mode match {
+      case SaveMode.SignUp => addUser(user.asInstanceOf[YetuUser])
+      case SaveMode.PasswordChange => {
+
+        for {
+          oldUser <- findUser(user.userId).map {
+            case Some(u) => {
+              val modifiedUser = u.copyUser(passwordInfo = user.passwordInfo)
+              usersIds += user.userId -> modifiedUser
+            }
+            case _ => Unit
+          }
+          find <- findUser(user.userId)
+        } yield find
+
       }
-      userToReturn
+      case _ =>
+        logger.warn("not saving as signUp; ignoring request.")
+        findUser(user.userId)
     }
+
+    result.map(_.orNull)
   }
 
-  def addNewUser(user: YetuUser) = {
-    val userWithRegistrationDate = user.copy(registrationDate = Some(new Date()))
-    users = users + (user.userId -> userWithRegistrationDate)
-    user
+  def addUser(user: YetuUser) = {
+
+    val savedUser = if (user.userAgreement.isDefined) {
+
+      user.registrationDate = Some(DateTime.now())
+
+      usersIds += (user.userId -> user)
```

Comment:
Is this inside an Actor or something that will handle concurrency for you? Otherwise if this code is called simultaneously from 2 Threads you might lose updates.